core/services/pipeline/orm.go

```diff
@@ -271,10 +271,41 @@ func (o *orm) InsertFinishedRun(run *Run, saveSuccessfulTaskRuns bool, qopts ...
 	return errors.Wrap(err, "InsertFinishedRun failed")
 }
 
+// DeleteRunsOlderThan deletes all pipeline_runs that have been finished for a certain threshold to free DB space
 func (o *orm) DeleteRunsOlderThan(ctx context.Context, threshold time.Duration) error {
 	q := o.q.WithOpts(pg.WithParentCtx(ctx))
-	err := q.ExecQ(`DELETE FROM pipeline_runs WHERE finished_at < $1`, time.Now().Add(-threshold))
-	return errors.Wrap(err, "DeleteRunsOlderThan failed")
+
+	err := pg.Batch(func(_, limit uint) (count uint, err error) {
+		result, cancel, err := q.ExecQIter(`
+WITH batched_pipeline_runs AS (
+	SELECT * FROM pipeline_runs
+	WHERE finished_at < ($1)
+	ORDER BY created_at ASC
```

Comment:
I would actually suggest `ORDER BY finished_at ASC` since this means it can do the whole thing with one index, it will be slightly more performant.

Reply:
Makes sense.